File: largestvalue_in_each_tree_row.java
//Time Complexity: O(N)
//Space Complexity: O(N) as there can be max n/2 nodes at the last tree in the case of an complete tree

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> largestValues(TreeNode root) {
        //edge case
        List<Integer> ans = new ArrayList<>();
        if(root == null)
            return ans;
        
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        
        while(!q.isEmpty())
        {
            int size = q.size();
            int max = Integer.MIN_VALUE;
            for(int i=0; i<size; i++)
            {
                TreeNode curr = q.poll();
                if(curr.val > max)
                    max = curr.val;
                if(curr.left != null) q.add(curr.left);
                if(curr.right != null) q.add(curr.right);
            }
            ans.add(max);
        }
        return ans;
    }
}

File: LargestValueInEachTreeRow.py
#TC:O(n)
#SC:O(n)
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def largestValues(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        result=[]
        q=[]
        q.append(root)
        if root==None : return result
        idx=0
        while q:
            l=len(q)
            
            for i in range(l):
                curr=q.pop(0)
                if idx==len(result):
                    result.append(curr.val)
                if result[idx]<curr.val:
                    result[idx]=curr.val
                
                if curr.right:
                    q.append(curr.right)
                if curr.left:
                    q.append(curr.left)
            
            idx+=1
        return result